### DISCUSSION TITLE
Feedback on Microsoft Pinyin Candidate List Issue after Windows 10 KB5043064 Update

### DISCUSSION THREAD

**Discussion by cary-rowen** [nvaccess/nvda#17223] [General]
**Issue:**  Installing Windows 10 update KB5043064 causes the candidate list in Microsoft Pinyin to stop refreshing correctly. This severely impacts the Chinese input experience for screen reader users, preventing them from knowing the real-time candidate character options.

This is a piece of feedback posted here to seek help from the community and would rather get a response from Microsoft.

I have reported this issue to Microsoft in Chinese.
See: https://aka.ms/AAsgszh

## Testing Environment:

Because Windows' built-in Narrator screen reader lacks proper support for Chinese input with Microsoft Pinyin, this report uses the NVDA screen reader ([https://github.com/nvaccess/nvda](https://github.com/nvaccess/nvda)) for demonstration.

## Steps to reproduce:

1. Download and install the NVDA screen reader: [https://www.nvaccess.org/files/nvda/releases/2024.3.1/nvda_2024.3.1.exe](https://www.nvaccess.org/files/nvda/releases/2024.3.1/nvda_2024.3.1.exe)
2. Download the IMEExpressive add-on for NVDA: [https://github.com/nvdacn/ime_expressive/releases/download/2024.3.21/ime_expressive-2024.3.21.nvda-addon](https://github.com/nvdacn/ime_expressive/releases/download/2024.3.21/ime_expressive-2024.3.21.nvda-addon)
3. While NVDA is running, double-click the `.nvda-addon` package to install the add-on. Follow the prompts to restart NVDA and complete the installation.
4. Open Notepad, switch the input method to Microsoft Pinyin, and type the keys "l", "u", "o" in sequence.

**Note:** The IMEExpressive Add-on used in this case offers highly customized features for Chinese input, such as providing word descriptions for the current candidate. This add-on is not related to the issue itself.

## Expected Behavior:

* When "l" is typed, NVDA should announce: "了解的了" 
* When "u" is typed, NVDA should announce: "记录的录"
* When "o" is typed, NVDA should announce: "包罗的罗"

**Note:** NVDA announces these phrases to provide word descriptions for the first candidate character in the input candidate list, making it easier for screen reader users to understand the current candidate. This is an enhancement provided by the IMEExpressive add-on.

## Actual Behavior:

* When "l" is typed, NVDA announces: "了解的了"
* When "u" is typed, NVDA has no voice feedback
* When "o" is typed, NVDA has no voice feedback

**Note:** This happens because NVDA only receives the candidate list update event after the first key is pressed.

**Note:** If you uninstall this update it works as expected.

## Preliminary Debugging Based on NVDA Source Code:**

Updating the input candidate list depends on the `handleInputCandidateListUpdate` function in NVDA: [https://github.com/nvaccess/nvda/blob/1573904b17722303f66434804cd515fc012b78af/source/NVDAHelper.py#L427](https://github.com/nvaccess/nvda/blob/1573904b17722303f66434804cd515fc012b78af/source/NVDAHelper.py#L427)

Each time a key is pressed (and the candidate list updates), this function should be called. This allows NVDA to get information about the current candidate character, its index in the list, and the input method being used.  However, adding debug statements to this function shows that it is only called once, even when multiple keys are pressed.

Although `handleInputCandidateListUpdate` is a higher-level Python wrapper, its core implementation resides in: [https://github.com/nvaccess/nvda/blob/master/nvdaHelper/remote/ime.cpp](https://github.com/nvaccess/nvda/blob/master/nvdaHelper/remote/ime.cpp)

### Key functions are implemented in the following locations:

* `handleCandidates`: [https://github.com/nvaccess/nvda/blob/master/nvdaHelper/remote/ime.cpp#L320](https://github.com/nvaccess/nvda/blob/master/nvdaHelper/remote/ime.cpp#L320)
* `handleIMEWindowMessage`: [https://github.com/nvaccess/nvda/blob/master/nvdaHelper/remote/ime.cpp#L455](https://github.com/nvaccess/nvda/blob/master/nvdaHelper/remote/ime.cpp#L455)
* `IME_getMessageHook`: [https://github.com/nvaccess/nvda/blob/master/nvdaHelper/remote/ime.cpp#L537](https://github.com/nvaccess/nvda/blob/master/nvdaHelper/remote/ime.cpp#L537)

**Comment by Adriani90** [comment]
cc: @msftedad

**Comment by cary-rowen** [comment]
cc @josephsl 
Can you provide some information? I was wondering, for Korean users, if a add-on is needed to input Korean, and would NVDA's current support for Korean input methods suffice?

**Comment by josephsl** [comment]
Hi,

There is no add-on specific to Korean IME (not that I know of). Also, try reproducing the issue WITHOUT the add-on and see if we can narrow this down further - NVDA issue, a recent change with IME interface, a combination of them, or neither.

Thanks.

**Comment by cary-rowen** [comment]
Thanks @josephsl 
The results are the same without the add-ons. I'm pretty sure it's a bug introduced by Windows Update.
